hbase-common/src/main/java/org/apache/hadoop/hbase/HConstants.java

```diff
@@ -1445,6 +1445,9 @@
       "hbase.util.default.lossycounting.errorrate";
   public static final String NOT_IMPLEMENTED = "Not implemented";
 
+  // Default Snapshot TTL - 30 days (good enough?)
+  public static final long DEFAULT_SNAPSHOT_TTL = 24 * 3600 * 30;
```

Comment:
So, what's the usage of the default TTL? If user doesn't specify TTL, should we just use default Snapshot API to take snapshot and keep it forever? If user specifies TTL, we use the value user set.

Reply:
If the user doesn't specify TTL, we can take above value as TTL by default. Otherwise we can take user provided TTL anyways. However, if the user decides to keep the snapshot forever(worst case), the user can specify negative value as TTL. e.g.: https://github.com/apache/hbase/pull/371/files#diff-f7c3e813539ba771a69ad518ef8dee00R2772

Does this look good to you?

Reply:
We should keep snapshot feature behaving the same as older version, especially if we are going to back-port branch-1 and branch-2, etc., i would suggest keeping snapshot forever by default.

Reply:
@Reidddddd Thanks for your review. Should we keep all previous snapshots forever? I thought may be 30 days could be good? IMO if the snapshot is kept for more than a month, it could be a real miss. What is your opinion? 
Moreover, I tried to provide one config **hbase.master.cleaner.snapshot.disable**, which if provided value "true", will prevent any snapshot from getting deleted - especially useful if some backup restore activity is going on for a couple of days.

cc: @apurtell

Reply:
If hbase.master.cleaner.snapshot.disable is false by default, then 30 days is ok, since it is explicitly set true.

Reply:
I agree with @Reidddddd that we should keep the old behavior and keep snapshots forever when TTL is not specified.

In case a user wants to have a new snapshot with TTL but have existing snapshots from a previous installation they will enable this feature and the old snapshots going to be deleted in 30 days. This shouldn't be the case.

Reply:
@petersomogyi @Reidddddd Thanks for your opinions. Can we have a new config? Say if **hbase.master.snapshot.apply.default.cleaner** is present with value "true", then any new snapshot created without TTL should have default TTL of 30 days applied. If the config is "false" or not present, and if user doesn't specify TTL, snapshot will be retained forever.
(default value of **hbase.master.snapshot.apply.default.cleaner** in default-site would be false)

I can incorporate this if you agree with this new config.

Reply:
Updated the PR. Please review

Reply:
What is the benefit of forcing TTL if a user does not specify it explicitly? 
Even at other areas where TTL can be used (e.g. Cell TTL) we do not force a default TTL value. In case that is not specified by the user HBase will keep those cells forever.

To keep backward compatibility (even if it is hidden behind a secondary configuration parameter) and consistency with other, similar features I'd prefer to have TTL=FOREVER as default value.

Reply:
Sure @petersomogyi I agree with having default TTL as Forever.
Would you be comfortable with this change: https://github.com/apache/hbase/pull/371/files#diff-f7c3e813539ba771a69ad518ef8dee00R2767 ?

Reply:
I definitely prefer this version rather than fixed 30 days TTL.

What do you think about having a config parameter e.g. `hbase.master.snapshot.ttl.default` with value 0 or -1 which means TTL FOREVER. Users can change it 30 days if that is the preferred value for their use case. With this we keep compatibility and users could specify default snapshot TTL based on their needs.

Reply:
The default TTL must be "forever". Anything else violates the principle of least surprise. 

Use 0 to denote "forever"